### PR TITLE
Show volume's usage in "Add disk" dialog as a Form helper text

### LINKED ISSUE #754: Add disk inline alert should be a contextual form error
This is a contextual, targeted error, but it's styled as an inline alert.

image

Inline alerts should only be used when the error is generic and at the top of a modal. It's best to have an error that's specific and located next to where the problem comes from (which this actually does) and styled as a contextual error (which this is not currently).

See: https://www.patternfly.org/v4/components/form/design-guidelines/#errors-and-validation

### PR DESCRIPTION
**Pull request by skobyda**
Instead of an inline dialog, an information about volume being used a
disk by another VM should be shown as a form helper text.

Fixes #754

### CODE REVIEW

**Comment by skobyda** [comment]
Alternatively we also discussed showing this as an error or a text variant. As there are use cases (e.g. disk sharing) where user might want to attach a VM used by another VM, therefore it should not be an error. I gave it some thought and I agree with what was mentioned on standup, that it's probably the best fitting as a warning variant, as disk sharing is quite a corner case.

Screenshot from 2022-07-12 02-00-33

**Comment by jelly** [comment]
Please rebase for the packit tests, the timeout is now 60 minutes.

**Review by garrett — APPROVED** [review]
Thanks! That's better.

**Review by garrett — CHANGES_REQUESTED** [review]
Ah, wait. Why isn't the select showing an underline and icon?

https://www.patternfly.org/v4/components/form/design-guidelines/#errors-and-validation

I think it's supposed to look like this:

image

https://www.patternfly.org/v4/components/select/html#warning

For React, set `validated` to `warning`.

**Comment by skobyda** [comment]
Fixed:
Screenshot from 2022-07-13 01-08-16

**Comment by skobyda** [comment]
Tests getting quite green, @martinpitt @jelly @marusak Mind giving it a code review?

**Review by garrett — DISMISSED** [review]
Looks good! Thanks!

Minor nitpick on the string; it shouldn't use a colon and period. Please remove both. Once this is done, it's good to go from my end.

image

It'd look like this:

image

**Comment by skobyda** [comment]
@garrett FYI There is one more situation where the warning contains 2 sentences. If the disk is raw, it's being automatically transfprmed to "shareable" disk, and the warning looks like this:
Screenshot from 2022-07-13 12-03-13

**Comment by skobyda** [comment]
Updated the PR.

### For non-raw disks, the warning looks like this:
Screenshot from 2022-07-13 12-07-16

### For raw disks, the warning looks like this:
Screenshot from 2022-07-13 12-06-53

### If there is multiple VMs using the disk, they are separated by comas:
Screenshot from 2022-07-13 12-09-01

File: src/components/vm/disks/diskAdd.jsx [hunk 1]
```diff
@@ -63,7 +63,22 @@ function getFilteredVolumes(vmStoragePool, disks) {
...
 
-const SelectExistingVolume = ({ idPrefix, storagePoolName, existingVolumeName, onValueChanged, vmStoragePools, vmDisks }) => {
+function getDiskUsageMessage(vms, storagePool, volumeName) {
+    const isVolumeUsed = getStorageVolumesUsage(vms, storagePool);
+
+    if (!isVolumeUsed[volumeName] || (isVolumeUsed[volumeName].length === 0))
+        return null;
+
+    const vmsUsing = isVolumeUsed[volumeName].join(', ');
+    let text = _("This volume is already used by ") + vmsUsing;
```
Comment (martinpitt) [inline src/components/vm/disks/diskAdd.jsx:73]: This doesn't work -- not every language can put the object at the very end. Please use "... by $0" and cockpit.format().

File: src/components/vm/disks/diskAdd.jsx [hunk 2]
```diff
@@ -63,7 +63,22 @@ function getFilteredVolumes(vmStoragePool, disks) {
...
+    const volume = storagePool.volumes.find(vol => vol.name === volumeName);
+    if (volume.format === "raw")
+        text += ". " + _("Attaching it will make this disk shareable for every VM using it.");
```
Comment (martinpitt) [inline src/components/vm/disks/diskAdd.jsx:76]: This kind of addition may also be problematic. It's not a given that every language uses a period. Honestly, I'd rather compromise a little and always add the period to the "already used by ..." message. Otherwise you need *two* string templates which just differ by ".", and that's a little rude towards translators. But this doesn't map grammar-wise very well. E.g. in German you would always end a sentence with a period. (Honestly it's rather confusing that English doesn't? or is that just some PF guideline? If so, it's very English centric.)

Alternatively, force a line break and drop the period from both sentences? Then there's at least some consistency in the UI.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Dakujem!

**Comment by garrett** [comment]
What does "Attaching it will make this disk shareable for every VM using it" mean? :thinking:

**Comment by skobyda** [comment]
> What does "Attaching it will make this disk shareable for every VM using it" mean? thinking

It will make it "Writeable and shareable"
Screenshot from 2022-07-14 12-34-30
https://access.redhat.com/documentation/en-us/red_hat_virtualization/4.1/html/administration_guide/shareable_disks_in_red_hat_enterprise_virtualization
The whole discussion why we do this can be found here: https://github.com/cockpit-project/cockpit/pull/12314#issuecomment-529871971

File: src/components/vm/disks/diskAdd.jsx (hunk 2 above)
Comment (skobyda) [reply src/components/vm/disks/diskAdd.jsx:76]: I like the line break idea. We use something similar for snapshots:

Comment (skobyda) [reply src/components/vm/disks/diskAdd.jsx:76]: On the second thought, it seems that pattenrfly forces one-liner text for Form helperText by ignoring `\n`. We could do some workaround, but maybe it's just simpler to use sentences with dots on the same line.

Comment (martinpitt) [reply src/components/vm/disks/diskAdd.jsx:76]: +1 on just using proper sentences, yes.

Comment (skobyda) [reply src/components/vm/disks/diskAdd.jsx]: Reverted back to using sentences ending with a comma. But let's wait if @garrett agrees.

File: src/components/vm/disks/diskAdd.jsx [hunk 3]
```diff
@@ -64,7 +64,23 @@ function getFilteredVolumes(vmStoragePool, disks) {
...
+function getDiskUsageMessage(vms, storagePool, volumeName) {
+    const isVolumeUsed = getStorageVolumesUsage(vms, storagePool);
+
+    if (!isVolumeUsed[volumeName] || (isVolumeUsed[volumeName].length === 0))
+        return null;
+
+    const vmsUsing = isVolumeUsed[volumeName].join(', ');
+    const volume = storagePool.volumes.find(vol => vol.name === volumeName);
+
+    let message = cockpit.format(_("This volume is already used by $0. "), vmsUsing);
```
Comment (martinpitt) [inline src/components/vm/disks/diskAdd.jsx:76]: Please no leading or trailing spaces, they are almost guaranteed to get lost in translation. Please add the " " separately (and not translated)

**Review by martinpitt — CHANGES_REQUESTED** [review]

**Review by martinpitt — APPROVED** [review]
Dakujem!

**Comment by garrett** [comment]
Oh, this is another one that was prematurely merged.

As I pointed out, the string "Attaching it will make this disk shareable for every VM using it" doesn't make sense. I couldn't suggest anything else as I didn't know what it meant (nor would our users), so I asked what it meant.

I still don't know:
- What "it" is
- If this will attach it if it's selected
- What "attach" means in this context of the "add" dialog (adding the disk will "attach" it right?)
- If this is a bad thing

This needs an immediate fixup to land before the release or it should be reverted.

File: src/components/vm/disks/diskAdd.jsx (hunk 2 above)
Comment (garrett) [reply src/components/vm/disks/diskAdd.jsx]: You didn't wait before merging.

**Comment by garrett** [comment]
Additionally, "Previously taken snapshots allow you to revert to an earlier state if something goes wrong"?!?!?

But there are no snapshots that have been previously taken. It's an empty state!

This also needs rewording.

**Comment by martinpitt** [comment]
> Additionally, "Previously taken snapshots allow you to revert to an earlier state if something goes wrong"?!?!?

This PR does not touch snapshots, so this needs to be changed independently. I'll split it out into an issue.
